libminifi/src/c2/C2Client.cpp

```diff
@@ -0,0 +1,318 @@
+        monitor->setStateMonitor(update_sink);
+      }
+      auto flowMonitor = std::dynamic_pointer_cast<state::response::FlowMonitor>(processor);
+      if (flowMonitor != nullptr) {
+        for (auto &con : connections) {
+          flowMonitor->addConnection(con.second);
+        }
+        flowMonitor->setStateMonitor(update_sink);
+        flowMonitor->setFlowVersion(flow_configuration_->getFlowVersion());
+      }
+      std::lock_guard<std::mutex> guard(metrics_mutex_);
+      root_response_nodes_[processor->getName()] = processor;
+    }
+  }
+
+  // get all component metrics
+  if (root_ != nullptr) {
+    std::vector<std::shared_ptr<core::Processor>> processors;
+    root_->getAllProcessors(processors);
+    for (const auto &processor : processors) {
+      auto rep = std::dynamic_pointer_cast<state::response::ResponseNodeSource>(processor);
+      // we have a metrics source.
+      if (nullptr != rep) {
+        std::vector<std::shared_ptr<state::response::ResponseNode>> metric_vector;
+        rep->getResponseNodes(metric_vector);
+        std::lock_guard<std::mutex> guard(metrics_mutex_);
+        for (auto& metric : metric_vector) {
+          component_metrics_[metric->getName()] = metric;
+        }
+      }
+    }
+  }
+
+  loadC2ResponseConfiguration("nifi.c2.root.class.definitions");
+
+  if (!initialized_) {
+    c2_agent_ = std::unique_ptr<c2::C2Agent>(new c2::C2Agent(controller, update_sink, configuration_, filesystem_));
+    c2_agent_->start();
+    initialized_ = true;
+  }
+}
+
+void C2Client::loadC2ResponseConfiguration(const std::string &prefix) {
+  std::string class_definitions;
+  if (!configuration_->get(prefix, class_definitions)) {
+    return;
+  }
+
+  std::vector<std::string> classes = utils::StringUtils::split(class_definitions, ",");
+
+  for (const std::string& metricsClass : classes) {
+    try {
+      std::string option = prefix + "." + metricsClass;
+      std::string classOption = option + ".classes";
+      std::string nameOption = option + ".name";
+
+      std::string name;
+      if (!configuration_->get(nameOption, name)) {
+        continue;
+      }
+      std::shared_ptr<state::response::ResponseNode> new_node = std::make_shared<state::response::ObjectNode>(name);
+      if (configuration_->get(classOption, class_definitions)) {
+        std::vector<std::string> classes = utils::StringUtils::split(class_definitions, ",");
+        for (const std::string& clazz : classes) {
+          // instantiate the object
+          auto ptr = core::ClassLoader::getDefaultClassLoader().instantiate(clazz, clazz);
+          if (nullptr == ptr) {
+            bool found_metric = [&] {
+              std::lock_guard<std::mutex> guard{metrics_mutex_};
+              auto metric = component_metrics_.find(clazz);
+              if (metric != component_metrics_.end()) {
+                ptr = metric->second;
+                return true;
+              }
+              return false;
+            }();
```

Comment:
Why is this a lambda expression?

Reply:
it is an IIFE, it initializes `found_metric` while doing some "complex" thing, in this case a search, and some side effect, it also limits the scope of the lock to the smallest extent possible

Reply:
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-lambda-init